test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -1,54 +1,133 @@
+ * @param check Callable for validating p
+ * @param non_random_generator_instance Callable for generating possible
+ * parameters
+ * @throws std::domain_error if non_random_generator_instance cannot satisfy
+ * check constraint
  */
-template <typename T1>
-void generateValues(T1& v, Constraint c, int N) {
-  v.resize(N);
-  
-  if (c == Constraint::None) {
-    for (int i = 0; i < N; i++)
-      v[i] = i - 1;
-  } else if (c == Constraint::Positive) {
-    for (int i = 0; i < N; i++)
-      v[i] = i + 1;
-  } else if (c == Constraint::NonNegative) {
-    for (int i = 0; i < N; i++)
-      v[i] = i;
+template<typename T1>
+struct generate_values {
+  template <typename T_check, typename T_generator>
+  T1 operator()(int N, T_check check,
+                T_generator &non_random_generator_instance) {
+    T1 vec(N);
+
+    for (int i = 0; i < vec.size(); i++) {
+      double v;
+
+      // There are only .size() possible non_random_numbers to try
+      for (int j = 0; j < non_random_generator_instance.size(); j++) {
+        v = non_random_generator_instance();
+        if (check(v))
+          break;
+      }
+
+      // If last number tried doesn't pass check, throw exception (constraint
+      // cannot be satisfied)
+      if (!check(v))
+        stan::math::domain_error("generate_values", "check", "",
+                                 "constraint cannot be satisfied by generator");
+
+      vec[i] = v;
+    }
+
+    return vec;
   }
-}
+};
 
 /*
- * Set v to a value that satisfies the constraint (and as much as possible
- * does not satisfy other constraints -- but this isn't totally possible)
+ * Generate a value from non_random_generator_instance that satisfies the
+ * constraint check.
+ *
+ * This is a functor instead of a function because it takes advantage of partial
+ * template specialization.
  *
  * @tparam T1 Type of v
- * @param v Variable to assign value to
- * @param c Constraint that value must fulfill
+ * @tparam T_check Callable type that takes in one double and returns a bool
+ * @tparam T_generator Callable type that takes no arguments and returns a
+ * double
  * @param N Unused
+ * @param check Callable for validating p
+ * @param non_random_generator_instance Callable for generating possible
+ * parameters
+ * @throws std::domain_error if non_random_generator_instance cannot satisfy
+ * check constraint
  */
-template <>
-void generateValues(double& v, Constraint c, int N) {
-  if (c == Constraint::None) {
-    v = -1;
-  } else if (c == Constraint::Positive) {
-    v = 1;
-  } else if (c == Constraint::NonNegative) {
-    v = 0;
+template<>
+struct generate_values<double> {
+  template <typename T_check, typename T_generator>
+  double operator()(int N, T_check check,
+                    T_generator &non_random_generator_instance) {
+    double v;
+
+    // There are only .size() possible non_random_numbers to try
+    for (int j = 0; j < non_random_generator_instance.size(); j++) {
```

Comment:
I think you can replace this loop with something like `none_of`: http://en.cppreference.com/w/cpp/algorithm/all_any_none_of